test/widgets/compose_box_test.dart

```diff
@@ -573,97 +600,199 @@ void main() {
+        required User user,
+        required UserRole role,
+      }) async {
+        await store.handleEvent(RealmUserUpdateEvent(id: 1,
+          userId: user.userId, role: role));
+        await tester.pump();
+      }
+
+      Future<void> changeChannelPolicy(WidgetTester tester, {
+        required ZulipStream channel,
+        required ChannelPostPolicy policy,
+      }) async {
+        await store.handleEvent(eg.channelUpdateEvent(channel,
+          property: ChannelPropertyName.channelPostPolicy, value: policy));
+        await tester.pump();
+      }
+
+      testWidgets('user role decreases -> compose box is replaced with the banner', (tester) async {
+        final selfUser = eg.user(role: UserRole.administrator);
+        await prepareComposeBox(tester,
+          narrow: const ChannelNarrow(1),
+          selfUser: selfUser,
+          streams: [eg.stream(streamId: 1, channelPostPolicy: ChannelPostPolicy.administrators)]);
         checkComposeBox(isShown: true);
+
+        await changeUserRole(tester, user: selfUser, role: UserRole.moderator);
+        checkComposeBox(isShown: false);
       });
-    });
 
-    group('group DMs', () {
-      testWidgets('compose box replaced with a banner', (tester) async {
-        final deactivatedUsers = [eg.user(isActive: false), eg.user(isActive: false)];
-        await prepareComposeBox(tester, narrow: groupDmNarrowWith(deactivatedUsers),
-          users: deactivatedUsers);
+      testWidgets('user role increases -> banner is replaced with the compose box', (tester) async {
+        final selfUser = eg.user(role: UserRole.guest);
+        await prepareComposeBox(tester,
+          narrow: const ChannelNarrow(1),
+          selfUser: selfUser,
+          streams: [eg.stream(streamId: 1, channelPostPolicy: ChannelPostPolicy.moderators)]);
         checkComposeBox(isShown: false);
+
+        await changeUserRole(tester, user: selfUser, role: UserRole.administrator);
+        checkComposeBox(isShown: true);
       });
 
-      testWidgets('at least one user becomes deactivated -> '
-          'compose box is replaced with a banner', (tester) async {
-        final activeUsers = [eg.user(isActive: true), eg.user(isActive: true)];
-        await prepareComposeBox(tester, narrow: groupDmNarrowWith(activeUsers),
-          users: activeUsers);
+      testWidgets('channel policy becomes stricter -> compose box is replaced with the banner', (tester) async {
+        final selfUser = eg.user(role: UserRole.guest);
+        final channel = eg.stream(streamId: 1, channelPostPolicy: ChannelPostPolicy.any);
+        await prepareComposeBox(tester,
+          narrow: const ChannelNarrow(1),
+          selfUser: selfUser,
+          streams: [channel]);
         checkComposeBox(isShown: true);
 
-        await changeUserStatus(tester, user: activeUsers[0], isActive: false);
+        await changeChannelPolicy(tester, channel: channel, policy: ChannelPostPolicy.fullMembers);
```

Comment:
nit: line too long — the key information is past 80 columns

(similarly the initial channelPostPolicy value above)